ipykernel/ipkernel.py

```diff
@@ -706,6 +724,58 @@ def do_clear(self):
             self.shell.reset(False)
         return dict(status="ok")
 
+    def _associate_identity_of_new_threads_with(self, parent_header):
+        """Intercept the identity of any thread started after this method finished,
+
+        and associate the thread's output with the parent header frame, which allows
+        to direct the outputs to the cell which started the thread.
+
+        This is a no-op if the `self._stdout` and `self._stderr` are not
+        sub-classes of `OutStream`.
+        """
+        stdout = self._stdout
+        stderr = self._stderr
+
+        def start_closure(self: threading.Thread):
+            """Wrap the `threading.Thread.start` to intercept thread identity.
+
+            This is needed because there is no "start" hook yet, but there
+            might be one in the future: https://bugs.python.org/issue14073
+            """
+
+            threading_start(self)
+            for stream in [stdout, stderr]:
+                if isinstance(stream, OutStream):
+                    stream._thread_parents[self.ident] = parent_header
+
+        threading.Thread.start = start_closure  # type:ignore[method-assign]
```

Comment:
I don't think it thread-safe. If a thread starts a new thread, that new thread will start outputting in the last executed cell.
I've taken a different approach in Solara, where I patch the ctor and run from the start:
https://github.com/widgetti/solara/blob/0ac4767a8c5f8c8b221cafe41fa9ac36270adbe8/solara/server/patch.py#L336

I think this approach might work better.

Reply:
To summarize the strategy, the ctor (`__init__`) would take a reference to the parent_header, and in run, that parent header is then set in a thread_local storage (in your case `_thread_parents`).

Reply:
> If a thread starts a new thread, that new thread will start outputting in the last executed cell.

Correct (there is no such a problem with asyncio side of things).

> I think this approach might work better.

Thank you for the link, I will take a look!

Reply:
Taking a closer look at solara, `current_context` takes the role of `_thread_parents`, right? This does not seem to differ that much. Your approach of patching earlier on initialization rather than startup has the advantage that it will work with:

```python
from threading import Thread
from time import sleep

def child_target():
    for i in range(iterations):
        print(i, end='', flush=True)
        sleep(interval)

def parent_target():
    thread = Thread(target=child_target)
    sleep(interval)
    thread.start()

Thread(target=parent_target).start()
```

but still not with:

```python
def parent_target():
    sleep(interval)
    Thread(target=child_target).start()

Thread(target=parent_target).start()
```

do I see this right?

Reply:
In the end my implementation converged to overriding the same methods as yours after all in https://github.com/ipython/ipykernel/pull/1186/commits/e1258deac3e0fce7515efa29e444a0c4889def04 :)